Make Stan implementations more modular

sethaxen opened the pull request:
This PR restructures the Stan implementations of targets and transforms to make them more modular. They are designed so that one can use Stan `#include` statements to combine targets and transforms.

Each target `<target>` is spread across the files:
- `targets/<target>/<target>_functions.stan`: implements the `_lpdf` functions for simplex and log-simplex
- `targets/<target>/<target>_data.stan`: implements the `data` block
- `targets/<target>/<target>_model_(log_)simplex`: implements the `model` block for the (log-)simplex

Each transform `<transform>` is spread across the files:
- `transforms/<transform>/<transform>_functions.stan`: implements the `_constrain_lp` functions for the simplex and log-simplex
- `transforms/<transform>/<transform>_parameters_(log_)simplex.stan`: implements the `parameters` and `transformed parameters` blocks for the (log-)simplex

The new tests check that all transforms and targets when combined can be compiled and sampled and that the log-density agrees with the JAX implementations. Some of the utilities in the tests will later be moved out for use in the analysis workflow.

sethaxen committed:
Rename target_densities to targets

sethaxen committed:
Merge dirichlet and log-dirichlet definitions

sethaxen committed:
Remove function block for easy inclusion

sethaxen committed:
Move log-Dirichlet to its own stanfunction

sethaxen committed:
Add model blocks file for Dirichlet

sethaxen committed:
Add multi-logit-normal implementation

Adapted from mjhajharia/transforms#74

sethaxen committed:
Make `log_` a prefix

sethaxen committed:
Fix density of multi-logit-normal

sethaxen committed:
Fix docstring

sethaxen committed:
Move transforms to blocks subdirectory

sethaxen committed:
Move ALR functions code to stanfunctions file

sethaxen committed:
Rename to stanfunctions extension

sethaxen committed:
Move ExpandedSoftmax functions to stanfunctions file

sethaxen committed:
Fix target densities

sethaxen committed:
Run stanc formatter on targets

sethaxen committed:
Run stanc formatter on transforms

sethaxen committed:
Make targets completely modular

sethaxen committed:
Rename transforms files

sethaxen committed:
Make transforms modular

sethaxen committed:
Remove outdated stan_models directory

sethaxen committed:
Strip trailing newlines

sethaxen committed:
Prefix targets/transforms with name

This is necessary for using Stan includes

sethaxen committed:
Run formatter

sethaxen committed:
Remove leftover blocks

sethaxen committed:
Fix log-Dirichlet normalization factor

sethaxen committed:
Update test for transforms

sethaxen committed:
Fix function file name

sethaxen committed:
Test also multi-logit-normal

sethaxen committed:
Merge branch 'main' into modular_stan

sethaxen committed:
Add data-only argument qualifiers for ILR's V matrix

sethaxen committed:
Merge branch 'main' into modular_stan

sethaxen committed:
Update ALR constructor call

sethaxen committed:
Test StanStickbreaking

bob-carpenter commented on transforms/ALR/ALR_functions.stan:
```diff
@@ -0,0 +1,19 @@
+vector inv_alr_log_simplex_constrain_lp(vector y) {
+  int N = rows(y) + 1;
+  real r = log1p_exp(log_sum_exp(y));
+  vector[N] log_x;
```
This can just be

```
vector[N] log_x = append_row(y - r, -r);
```

I'd just skip defining `log_x` and substitute the expression above directly into the return.

bob-carpenter commented on transforms/ILR/ILR_functions.stan:
```diff
@@ -0,0 +1,29 @@
+  real inv_nrm2;
+  for (n in 1 : (N - 1)) {
+    inv_nrm2 = inv_sqrt(n * (n + 1));
+    V[1 : n, n] = rep_vector(inv_nrm2, n);
```
`V[ , n]` can also be defined by appending the vectors, but it's more efficient and arguably more readable as you wrote it.

bob-carpenter commented on transforms/StickbreakingAngular/StickbreakingAngular_functions.stan:
```diff
@@ -0,0 +1,46 @@
+  real log_halfpi = log(pi()) - log2();
+  int rcounter = 2 * N - 3;
+  for (i in 1 : (N - 1)) {
+    u = log_inv_logit(y[i]);
```
For `real`, it's conventional and just as efficient to declare and define at the same time.  The general consensus is that it's easier to read code if the declarations are as near the definitions as possible.

So that'd be

```stan
   real u = log_inv_logit(y[I]);
```

and so on.   With vectors, it can be more efficient to declare them once if there's going to be reallocation otherwise.

I would just go with whatever you find clearest here, but it's generally more efficient to vectorize the transforms, so that'd be

```stan
vector[N] log_phi = log_inv_logit(y) + (log(pi()) - log2());
```

and then replace uses of `log_phi` later with `log_phi[n]`.  The reason is that it reduces the size of the expression graph for autodiff.

sethaxen replied:
@bob-carpenter here's another example where I can completely eliminate the for loop for the log-simplex transform but with more vector allocations. Would this generally be more efficient in Stan?

```stan
vector stickbricking_angular_log_simplex_constrain_lp(vector y) {
  int N = rows(y) + 1;
  vector[N - 1] log_u = log_inv_logit(y);
  vector[N - 1] log_phi = log_u + log(pi()) - log2();
  vector[N - 1] phi = exp(log_phi);
  vector[N - 1] log_s = log(sin(phi));
  vector[N - 1] log_c = log(cos(phi));
  vector[N] log_s2_prod = cumulative_sum(append_row(0, 2 * log_s));
  vector[N] log_x = log_s2_prod + append_row(2 * log_c, 0);
  target += (N - 1) * log2();
  target += log_s2_prod;
  target += log_c;
  target += log_s;
  target += log_phi;
  target += log1m_exp(log_u);
  target += -log_x;
  return log_x;
}
```

bob-carpenter replied:
There's not a one-size-fits-all answer, I'm afraid.  It will generally depend on how big the vectors are, what other pressure there is on memory, and the amount of reduction in the expression graph and number of virtual function calls with vectorizations.

For example, this:

```stan
vector[N - 1] log_u = log_inv_logit(y);
```

will allocate the vector, but it will also reduce the number of virtual function calls to 1.  Whether that's a win will depend on compiler memory pressure, etc.  

I've rewritten the next line because arithmetic in C++ and hence Stan is evaluated left to right.  So it's better to combine scalars and then add to a vector.   

```stan
  vector[N - 1] log_phi = log_u + (log(pi()) - log2());
```

This should be faster because it simplifies the autodiff from a bunch of individual additions to vector addition, which simplifies the expression graph.  

The following

```stan
  vector[N - 1] phi = exp(log_phi);
```

is like the first expression, where the speedup or slowdown is going to depend on a lot of contingencies.

This

```stan
  vector[N] log_s2_prod = cumulative_sum(append_row(0, 2 * log_s));
```

is going to be a clear win over a loop because there's an efficient underlying gradient (I think---you'd really need to try it or look at the C++ code).  Now it'd probably be even faster to do the append_row on the outside. 

Then it depends on how they get used.  While creating the value might not be faster,

```stan
  target += -log_x;
```
is going to be a lot faster than doing it in a loop.

In general,

```stan
target += a;
target += b;
```

is faster than 

```stan
target += (a + b);
```

because under the hood, `target +=` just accumulates a list, which is fed into a final summation, which is more efficient than building up a binary tree of additions.

bob-carpenter commented on transforms/StickbreakingLogistic/StickbreakingLogistic_functions.stan:
```diff
@@ -0,0 +1,31 @@
+  real log_z;
+  real log_cum_prod = 0;
+  for (i in 1 : (N - 1)) {
+    log_z = log_inv_logit(y[i] - log(N - i)); // logistic_lcdf(y[i] | log(N - i), 1)
```
The built-in forms are usually more efficient as they produce more compact expression graphs for autodiff, but again ,no big deal here if you think this is clearer.

There's also a `cumulative_sum()` built-in, which could be used here.   It's also more efficient than the loop for the same reason---smaller autodiff graph and fewer virtual function calls in the backward pass.  Again, no big deal if you like it this way better.

sethaxen replied:
Ah, this will hopefully clear up some confusion I have about whether it's better in Stan to vectorize even if it means more array allocations. e.g. here `logistic_lcdf` is not vectorized, so to use that we would definitely define `log_z` inside the loop. But we could also eliminate the loop entirely and do this

```stan
  // vectorized logistic_lcdf(y[i] | log(N - i), 1)
  vector[N - 1] log_z = log_inv_logit(y - log(reverse(linspaced_vector(N - 1, 1, N - 1))));
  vector[N] log_x = append_row(log_z, 0) + cumulative_sum(append_row(0, log1m_exp(log_z)));
  target += log_x[N];
```

Is this significantly more efficient? If so, it might be preferable, even though IMO it is less readable.

bob-carpenter replied:
Right---if `logistic_lcdf` isn't itself vectorized, then it's tough.

Is that `+= log_x[N]` correct?  If so, it's best to just calculate the single value corresponding to `log_x[N], which I think is just `sum(log1m_exp(log_z))`.

sethaxen replied:
> Is that `+= log_x[N]` correct?

Yes.

> If so, it's best to just calculate the single value corresponding to `log_x[N], which I think is just `sum(log1m_exp(log_z))`.

I'm not sure I follow. You mean like this?

```stan
  vector[N - 1] log_z = log_inv_logit(y - log(reverse(linspaced_vector(N - 1, 1, N - 1))));
  vector[N] log_x = append_row(log_z, 0) + cumulative_sum(append_row(0, log1m_exp(log_z)));
  target += sum(log1m_exp(log_z));
```

But doesn't this end up duplicating the work we did to compute `log_x[N]`?

bob-carpenter replied:
Ah---I didn't see `log_x` being used in the target and didn't see that you returned it.  So no, forget my suggestion.

bob-carpenter commented on transforms/StickbreakingPowerLogistic/StickbreakingPowerLogistic_functions.stan:
```diff
@@ -0,0 +1,36 @@
+    log_w = log_u / (N - i);
+    log_z = log1m_exp(log_w);
+    x[i] = exp(log_cum_prod + log_z);
+    target += 2 * log_u - y[i]; // logistic_lupdf(y[i] | 0, 1);
```
This isn't necessarily going to be equal to the `logistic_lupdf`.  All the `_lupdf` function guarantees is equality up to an additive constant---we don't say exactly what's going to get dropped.

sethaxen replied:
Ah, thanks, that's actually an incorrect comment (should be `_lpdf`)

bob-carpenter commented on transforms/StickbreakingPowerLogistic/StickbreakingPowerLogistic_functions.stan:
```diff
@@ -0,0 +1,36 @@
+    log_z = log1m_exp(log_w);
+    log_x[i] = log_cum_prod + log_z;
+    target += 2 * log_u - y[i]; // logistic_lupdf(y[i] | 0, 1);
+    target += -log_x[i];
```
It's more efficient to pull this out of the loop and just use the vectorized `target +=`:

```stan
target += -log_x;
```

sethaxen replied:
Ah, okay, so `-log_x` doesn't end up allocating a new vector, or that doesn't really matter?
What about if I need to add the contents of several vectors `a` and `b` of the same length to the target. Is it better to do `target += a; target += b` or `target += a + b`, or are both the same?

bob-carpenter replied:
Actually, `-log_x` might allocate a new vector unless we're returning an expression graph.  Steve Bronder is going through and replacing the return types with expression templates, so that things like `-log_x` will not allocate a new vector.  When that's all done, vectorization will pretty much always be the right thing to do.

One issue with optimization is that as the compiler changes, so do the optimization techniques.  For example, ll the careful work I'd done on Java 1.4 optimization in the early 00s went out the window when they built most of it into the Java 1.5 compiler.  

I answered the `target +=` question above---better to split it.

bob-carpenter reviewed (approved):
Thanks so much!  This is some of the most beautiful Stan coding I've ever seen!  I made a few comments but it's really fine the way it is already, so I'll leave it up to you as to what you want to change and just approve.

sethaxen committed:
Use append_row in ALR

sethaxen committed:
Declare real variables only where defined

sethaxen committed:
Set seed for sampling and increase sigfigs

sethaxen committed:
Add docstring to log_dirichlet_lpdf

sethaxen committed:
Make log_dirichlet_lpdf define the return while returning

For consistency with the multi-logit-normal variants

sethaxen committed:
Implicitly increment target outside of loops

sethaxen committed:
Fix comment

sethaxen committed:
Retain all sig-figs, and choose new stan seed

sethaxen commented:
Of the suggestions, I made the following changes:
- Declared real variables only where defined
- When possible without allocating, increment the target outside of the loop

Because this PR is focused on reorganization of existing Stan code, I'll merge this PR as is and maybe in a future commit replace some loops with vectorized code.

sethaxen committed:
Concentrate draws towards middle of the simplex

This is where transforms are most likely to be numerically stable, so the small numerical differences between jax and stan transforms less likely to be relevant.
